### ISSUE TITLE
Wrong statement generated when child classes have same propery name but different type.

### ISSUE THREAD

**Issue by FelicePollano** [nhibernate/nhibernate-core#3363]
I think there is something non working while generating query in a hierarchy of classes in which some derived class have same property with different many-to-one class. ie:

                ```
                             +----------+
				 +  mother  +
				 +----------+
				 
				 
	+-------------+				+-------------+
	+    Child1   +				+    Child1   +
	===============				===============
	| Thing:Thing1|				| Thing:Thing2|
	+-------------+				+-------------+
```
So if we look for a specific value of Thing property like this:
`var result = session.Query<Mother>().Where(k => k is Child1 && (k as Child1).Thing.Id == "00001").ToList();`

We dont find record even if recors exists, as the query generated is left-jiuning the wrong table, ie:
```
select mother0_.Id as id1_0_, mother0_.Name as name3_0_,
					mother0_.thingId as thingid4_0_, mother0_.kind as kind2_0_ from Mother mother0_ 
					left outer join Thing2 thing2x1_ on 
					mother0_.thingId=thing2x1_.Id where mother0_.kind='1' and thing2x1_.Id=?"
```

that is, looks for able for Thing2, while we actually look for a Thing1...

I'm providing unit test as a PR.

**Comment by bahusoid** [comment]
Looks like duplicate of #1189

**Comment by FelicePollano** [comment]
Actually can be same issue. I can add I had the case working in production until 5.2.7, it broke after 5.3.0 included.

**Comment by fredericDelaporte** [comment]
So, it would be a 5.3.x regression. I consider checking this before releasing the next 5.4.x.

**Comment by FelicePollano** [comment]
> I have no good fix for this case in mind - you better avoid using the same names in subclasses for different entities.

actually this is current solution adopted, but, if this is the case, some excpetion in mapping phase should complain, otherwie is easy to be confused.

**Comment by FelicePollano** [comment]
By looking at current code base, looks like there is something wrong here: ( `AbstractPropertyMapping `)

```
protected void AddPropertyPath(string path, IType type, string[] columns, string[] formulaTemplates)
{
			typesByPropertyPath[path] = type;
			columnsByPropertyPath[path] = columns;

			if (formulaTemplates != null)
				formulaTemplatesByPropertyPath[path] = formulaTemplates;
}
```
So this function is called collecting a property name/type map for the **entire hierarchy of an entity**, so, if down the entire tree some property have same name, entry will be overwritten, causing unexpected behaviors. I had issues compiling 5.2.7 tag because of missing requirement, but maybe in the previous version this was called only for single entity?

**Comment by fredericDelaporte** [comment]
No, I do not think this has changed. What has changed is #2081, fixing #1274.

Your mapping use `not-found="ignore"` on the relation, which means a not null foreign key does not guarantee the associated entity exists. Prior to #2081, in case of comparison on the associated ID, NHibernate would do that on the foreign key without checking the associated entity actually exist, which could yield wrong results in case the entity was not actually existing.
Now it checks the associated entity existence, but is hit by #1189 in case of polymorphism.

(About issues for compiling, switching from 5.4.x to a lower version branch usually requires cleaning, closing the solution in VS, restoring. Actually, it is more reliable to close the solution after switching branch, purge all bin and obj folders, reopen the solution.)

So, by fixing a `not-found="ignore"` bug in some cases, #2081 has caused another `not-found="ignore"` bug (#1189) to be extended to some more cases. By the way, your application was surely hit by the bug fixed by #2081, unless it does not actually have missing entities. If it does not have missing entities, remove `not-found="ignore"`. It will solve the trouble.

**Comment by fredericDelaporte** [comment]
So, this is still a regression caused by #2081. But maybe a too tough one to fix, hitting what is in my opinion a corner case (`not-found="ignore"` + polymorphism with different association properties sharing the same name). Not sure we should mandate a fix for this one in 5.3.x.

**Comment by bahusoid** [comment]
> Not sure we should mandate a fix for this one in 5.3.x.

Agree. Not easy fixable. 

I see that in the test case the same column name `thingid` is used to reference two different tables (`Thing2` and `Thing1`). If `not-found="ignore"` is added just to avoid foreign key creation - you should use `foreing-key="none"` mapping instead.

**Comment by FelicePollano** [comment]
hi @bahusoid, no, it is not to avoid foreign key creation. By the way, even if not modified, from what I saw, the function `AddPropertyPath` mentioned above, looks like it is the source of the problem, in any hierarchy, i two property share same name, and that could actially happen for any reason, the type map is overvritten by the last entry found. Here the problem. I suspect that can be causing some issues somwhere else as well. Short "fix", according to me, could be at least throw an exception when this happens, pointing to the undesired duplicated proprty name.

**Comment by fredericDelaporte** [comment]
Yes the issue is more general (and known since long), but complex to fix, see #1189.

Still, your specific case could be fixed by a "work-around": cease using that other "work-around" that is `not-found="ignore"`.

(That `ignore` feature is a work-around for supporting incoherent data in the database. It is best to not use it if possible. This means, fix invalid data and enforce in the DB foreign key constraints to avoid new invalid data, then remove from the mapping these `ignore` settings.)
